modules/standard/Maps.chpl

```diff
@@ -0,0 +1,291 @@
+ */
+
+/*
+  This module contains the implementation of the map type which is a container
+  that stores key-value associations. 
+
+  The following operations may invalidate references to elements contained in
+  a map:
+
+      - indexing using a non-existing key (e.g. `myMap[key] = value`)
+      - add
+      - remove
+      - clear
+      - update
+
+  Maps are not parallel safe by default, but can be made parallel safe by
+  setting the param formal `parSafe` to true in any map constructor. When
+  constructed from another map, the new map will inherit the parallel safety
+  mode of its originating map.
+*/
+module Maps {
+
+  record map {
+
+    /*
+      Initializes an empty map containing keys and values of given types.
+
+      :arg keyType: The type of the keys of this map.
+      :arg valType: The type of the values of this map.
+      :arg parSafe: If `true`, this map will use parallel safe operations.
+    */
+    proc init(type keyType, type valType, param parSafe=false) {
+
+    }
+
+    /*
+      Initializes a map containing elements that are copy initialized from
+      the elements contained in another map.
+
+      :arg other: The map to initialize from.
+      :type other: map
+
+      :arg parSafe: If `true`, this map will use parallel safe operations.
+      :type parSafe: bool
+    */
+    proc init=(const ref other: map(?kt, ?vt), param parSafe=other.parSafe) {
+
+    }
+
+    /*
+      Clears the contents of this map.
+
+      .. warning::
+
+        Clearing the contents of this map will invalidate all existing
+        references to the elements contained in this map.
+    */
+    proc clear() {
+
+    }
+
+    /*
+      The current number of keys contained in this map.
+    */
+    inline proc const size {
+
+    }
+
+    /*
+      Returns `true` if this map contains zero keys.
+
+      :returns: `true` if this map is empty.
+      :rtype: `bool`
+    */
+    inline proc const isEmpty(): bool {
+
+    }
+
+    /*
+      Returns `true` if the given key is a member of this map, and `false`
+      otherwise.
+
+      :arg k: The key to test for membership.
+      :type k: keyType
+
+      :returns: Whether or not the given key is a member of this map.
+      :rtype: `bool`
+    */
+    proc const contains(const ref k: keyType): bool {
+
+    }
+
+    /*
+      Updates this map with the contents of the other, overwritting the values
+      for already-existing keys.
+
+      :arg m: The other map
+      :type m: map(keyType, valType)
+    */
+    proc update(const ref m: zap(keyType,valType)) {
+
+    }
+
+    /*
+      Get a reference to the value mapped to the given key.
+
+      :arg k: The key to access
+      :type k: keyType
+
+      .. warning::
+
+        Use of the `this` method with a non-existent key (while bounds checking
+        is on) will cause the currently running program to halt.
+
+      :returns: Reference to the value mapped to the given key.
+    */
+    proc const this(k: keyType) ref {
```

Comment:
So there's two ways to go about the proper design of indexing here:

- It will halt if there is an attempt to access a key not yet in the table
- It will emplace the key with a default initialized value if no pairing exists

I believe that the current status quo in Chapel is to do the latter, as it is the only way to enable idioms such as:

```chapel
someMap['foo'] = 4;
```

The downside of this approach is obviously the penalty of a pointless default initialization in the above assignment.

I'm still of two minds about this, however I think that ultimately I would prefer this to be a synonym for `addIfAbsent` as I think it enables behavior expected by a lot of users (I believe this is much why C++ opts to do the same). 

Since our current trend is to mimic Python insofar as is feasible, I think it's probably best to enable the above behavior instead.

Reply:
I agree that `someMap['foo'] = 4;` is something we should support without the need for an explicit `add`

But I am not sure those two are our only options here. We can probably let the compiler help us by implementing three overloads of `this` with non-ref, ref and const-ref return intents. Then, in your example, `ref` version would be called and then we can take action for insertion of a new value or updating an existing one. For others we can still do an OOB check. The best example for that would be the `dsiAccess` overloads in `modules/internal/DefaultAssociative`

Will update the doc shortly.

Reply:
I think regardless of the return type of `this`, you're still going to have to default initialize the value in an unused slot if the key hashes to it, with this method.

However thinking about this, I'm not even sure that this is something you can avoid with a method like `add(key, value), because assignment semantics are only defined when the LHS is in a consistent state (default initialized at some point in time). 

Extra compiler support to avoid returning a ref to primitives on the RHS in the situation where you have say, a `map(int, int)` is good though, and I'm all for that.

Reply:
@dlongnecke-cray - I'm just really confused by your comment. To the extent I'm able to respond at all... I lean towards continuing to have the behavior of associative arrays now (which we are trying to replace).

Reply:
The comment was with regards to being able to emplace a `(key, value)` when a key hashes onto an unused slot in the map while avoiding the cost of default initializing the _value_ before assigning it.

A method like `add(in k, in v)` could do this with a move while `foo['x'] = 5;` would have to default initialize the slot that 5 is assigned to first.

So just a bit of extra potential work (?).